pageserver/src/storage_sync/upload.rs

```diff
@@ -1,479 +1,110 @@
+use fail::fail_point;
+use std::path::Path;
 use tokio::fs;
-use tracing::{debug, error, info, warn};
-
-use utils::id::TenantTimelineId;
 
-use super::{
-    index::{IndexPart, RemoteTimeline},
-    LayersUpload, SyncData, SyncQueue,
-};
-use crate::metrics::NO_LAYERS_UPLOAD;
-use crate::{config::PageServerConf, storage_sync::SyncTask};
+use super::index::IndexPart;
+use crate::config::PageServerConf;
+use crate::storage_sync::LayerFileMetadata;
+use remote_storage::GenericRemoteStorage;
+use utils::id::{TenantId, TimelineId};
 
 /// Serializes and uploads the given index part data to the remote storage.
-pub(super) async fn upload_index_part(
+pub(super) async fn upload_index_part<'a>(
     conf: &'static PageServerConf,
-    storage: &GenericRemoteStorage,
-    sync_id: TenantTimelineId,
-    index_part: IndexPart,
+    storage: &'a GenericRemoteStorage,
+    tenant_id: TenantId,
+    timeline_id: TimelineId,
+    index_part: &'a IndexPart,
 ) -> anyhow::Result<()> {
-    let index_part_bytes = serde_json::to_vec(&index_part)
-        .context("Failed to serialize index part file into bytes")?;
-    let index_part_size = index_part_bytes.len();
-    let index_part_bytes = tokio::io::BufReader::new(std::io::Cursor::new(index_part_bytes));
-
-    let index_part_path = conf
-        .metadata_path(sync_id.timeline_id, sync_id.tenant_id)
-        .with_file_name(IndexPart::FILE_NAME);
-    storage
-        .upload_storage_object(
-            Box::new(index_part_bytes),
-            index_part_size,
-            &index_part_path,
-        )
-        .await
-        .with_context(|| format!("Failed to upload index part for '{sync_id}'"))
-}
-
-/// Timeline upload result, with extra data, needed for uploading.
-#[derive(Debug)]
-pub(super) enum UploadedTimeline {
-    /// Upload failed due to some error, the upload task is rescheduled for another retry.
-    FailedAndRescheduled(anyhow::Error),
-    /// No issues happened during the upload, all task files were put into the remote storage.
-    Successful(SyncData<LayersUpload>),
+    fail_point!("before-upload-index", |_| {
+        bail!("failpoint before-upload-index")
+    });
+    async {
+        let index_part_bytes = serde_json::to_vec(&index_part)
+            .context("Failed to serialize index part file into bytes")?;
+        let index_part_size = index_part_bytes.len();
+        let index_part_bytes = tokio::io::BufReader::new(std::io::Cursor::new(index_part_bytes));
+
+        let index_part_path = conf
+            .metadata_path(timeline_id, tenant_id)
+            .with_file_name(IndexPart::FILE_NAME);
+        storage
+            .upload_storage_object(
+                Box::new(index_part_bytes),
+                index_part_size,
+                &index_part_path,
+            )
+            .await
+            .with_context(|| {
+                format!("Failed to upload index part for '{tenant_id} / {timeline_id}'")
+            })
+    }
+    .await
```

Comment:
Does the need for these inner `async { }.await` come from the `fail_point!` usage earlier or is there some `.instrument(info_span!(...))` missing which was the original idea?

Similarly in other fn(s).

Reply:
I think there's a `.measure_remote_op` call missing here, in upload_timeline_layer, and in delete_layer. I'll add them

Reply:
No, wait, we have them in the callers now. So these async blocks are pointless now. I'll remove the async blocks

Reply:
Opened https://github.com/neondatabase/neon/pull/2925 to clean that up.